src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateSchedulerActor.scala

```diff
@@ -2,33 +2,80 @@ package io.iohk.ethereum.blockchain.sync
 
 import akka.actor.{Actor, ActorLogging, ActorRef, Props, Timers}
 import akka.util.ByteString
+import cats.data.NonEmptyList
 import io.iohk.ethereum.blockchain.sync.LoadableBloomFilter.BloomFilterLoadingResult
-import io.iohk.ethereum.blockchain.sync.SyncStateDownloaderActor.{CancelDownload, RegisterScheduler}
-import io.iohk.ethereum.blockchain.sync.SyncStateScheduler.{ProcessingStatistics, SchedulerState, SyncResponse}
-import io.iohk.ethereum.blockchain.sync.SyncStateSchedulerActor.{
-  BloomFilterResult,
-  GetMissingNodes,
-  MissingNodes,
-  PrintInfo,
-  PrintInfoKey,
-  RestartRequested,
-  StartSyncingTo,
-  StateSyncFinished,
-  SyncStateSchedulerActorCommand,
-  StateSyncStats,
-  WaitingForNewTargetBlock
+import io.iohk.ethereum.blockchain.sync.PeerRequestHandler.ResponseReceived
+import io.iohk.ethereum.blockchain.sync.SyncStateScheduler.{
+  CriticalError,
+  ProcessingStatistics,
+  SchedulerState,
+  SyncResponse
 }
+import io.iohk.ethereum.blockchain.sync.SyncStateSchedulerActor._
+import io.iohk.ethereum.crypto.kec256
+import io.iohk.ethereum.network.p2p.messages.PV63.{GetNodeData, NodeData}
+import io.iohk.ethereum.network.{Peer, PeerId}
 import io.iohk.ethereum.utils.ByteStringUtils
 import io.iohk.ethereum.utils.Config.SyncConfig
 import monix.execution.Scheduler
 
+import scala.annotation.tailrec
+import scala.collection.immutable.Queue
+import scala.concurrent.Future
 import scala.concurrent.duration._
 
-class SyncStateSchedulerActor(downloader: ActorRef, sync: SyncStateScheduler, syncConfig: SyncConfig)
-    extends Actor
+class SyncStateSchedulerActor(
+    sync: SyncStateScheduler,
+    val syncConfig: SyncConfig,
+    val etcPeerManager: ActorRef,
+    val peerEventBus: ActorRef,
+    val scheduler: akka.actor.Scheduler
+) extends Actor
+    with PeerListSupport
+    with BlacklistSupport
     with ActorLogging
     with Timers {
-  implicit val scheduler = Scheduler(context.dispatcher)
+
+  implicit val monixScheduler = Scheduler(context.dispatcher)
+
+  def handleCommonMessages: Receive = handlePeerListMessages orElse handleBlacklistMessages
+
+  private def getFreePeers(state: DownloaderState) = {
+    handshakedPeers.collect {
+      case (peer, _) if !state.activeRequests.contains(peer.id) && !isBlacklisted(peer.id) => peer
+    }
+  }
+
+  private def requestNodes(request: PeerRequest): ActorRef = {
+    log.info(s"Requesting ${request.nodes.size} from peer ${request.peer}")
+    implicit val s = scheduler
+    val handler = context.actorOf(
+      PeerRequestHandler.props[GetNodeData, NodeData](
```

Comment:
Does it make sense to extract a function wrapping that actor creation and errors handling into a `Future`? Basically something merging `requestNodes` and `handleRequestResults` into a function like
```
def PeerRequesthandler.makeRequest[Req, Res](peer, message: Req): Future[Either[RequestFailure|Timeout, Res]] = {}
```

Reply:
imo not,  as we will probably end with something weird as `self ? Request` when we will need to specify artificial timeout event hough we would be asking self. Current way is imo more natural, when the requesting actor is creating `PeerRequestHandler` by itself. 

This approach would definitely be good improvement if we would introduce `PeerClient` and then ask it for those requests, but i would do it in separate task which would do the same also for main `FastSync` (it would great step into hiding `makeRequest...` behind some `Network` interface like in the other project)